RecoVertex/BeamSpotProducer/plugins/BeamSpotOnlineProducer.cc

```diff
@@ -48,94 +56,144 @@ BeamSpotOnlineProducer::BeamSpotOnlineProducer(const ParameterSet& iconf)
+  iDesc.addWithDefaultLabel(ps);
+}
+
 void BeamSpotOnlineProducer::produce(Event& iEvent, const EventSetup& iSetup) {
+  // product is a reco::BeamSpot object
+  auto result = std::make_unique<reco::BeamSpot>();
+  reco::BeamSpot aSpot;
   //shout MODE only in stable beam
   bool shoutMODE = false;
   edm::Handle<L1GlobalTriggerEvmReadoutRecord> gtEvmReadoutRecord;
   if (iEvent.getByToken(l1GtEvmReadoutRecordToken_, gtEvmReadoutRecord)) {
     if (gtEvmReadoutRecord->gtfeWord().beamMode() == theBeamShoutMode)
       shoutMODE = true;
   } else {
-    shoutMODE = true;
+      shoutMODE = true;
   }
-
-  // get scalar collection
-  Handle<BeamSpotOnlineCollection> handleScaler;
-  iEvent.getByToken(scalerToken_, handleScaler);
-
-  // beam spot scalar object
-  BeamSpotOnline spotOnline;
-
-  // product is a reco::BeamSpot object
-  auto result = std::make_unique<reco::BeamSpot>();
-
-  reco::BeamSpot aSpot;
-
   bool fallBackToDB = false;
-  if (!handleScaler->empty()) {
-    // get one element
-    spotOnline = *(handleScaler->begin());
-
-    // in case we need to switch to LHC reference frame
-    // ignore for the moment rotations, and translations
-    double f = 1.;
-    if (changeFrame_)
-      f = -1.;
-
-    reco::BeamSpot::Point apoint(f * spotOnline.x(), spotOnline.y(), f * spotOnline.z());
-
-    reco::BeamSpot::CovarianceMatrix matrix;
-    matrix(0, 0) = spotOnline.err_x() * spotOnline.err_x();
-    matrix(1, 1) = spotOnline.err_y() * spotOnline.err_y();
-    matrix(2, 2) = spotOnline.err_z() * spotOnline.err_z();
-    matrix(3, 3) = spotOnline.err_sigma_z() * spotOnline.err_sigma_z();
-
-    double sigmaZ = spotOnline.sigma_z();
-    if (theSetSigmaZ > 0)
-      sigmaZ = theSetSigmaZ;
-
-    aSpot = reco::BeamSpot(apoint, sigmaZ, spotOnline.dxdz(), f * spotOnline.dydz(), spotOnline.width_x(), matrix);
-
-    aSpot.setBeamWidthY(spotOnline.width_y());
-    aSpot.setEmittanceX(0.);
-    aSpot.setEmittanceY(0.);
-    aSpot.setbetaStar(0.);
-    aSpot.setType(reco::BeamSpot::LHC);  // flag value from scalars
-
-    // check if we have a valid beam spot fit result from online DQM
-    if (spotOnline.x() == 0 && spotOnline.y() == 0 && spotOnline.z() == 0 && spotOnline.width_x() == 0 &&
-        spotOnline.width_y() == 0) {
-      if (shoutMODE) {
-        edm::LogWarning("BeamSpotFromDB")
-            << "Online Beam Spot producer falls back to DB value because the scaler values are zero ";
+  if (useTransientRecord_){
+      auto const& spotDB = iSetup.getData(beamTransientToken_);  
+
+      if (spotDB.GetBeamType() != 2){            
+       if (shoutMODE) {
+          edm::LogWarning("BeamSpotFromDB")
+              << "Online Beam Spot producer falls back to DB value because the ESProducer returned a fake beamspot ";
+        }
+        fallBackToDB = true;
+      }else{
+        // translate from BeamSpotObjects to reco::BeamSpot
+        reco::BeamSpot::Point apoint(spotDB.GetX(), spotDB.GetY(), spotDB.GetZ());
```

Comment:
shouldn't also this mode support the frame conversion?
See:
https://github.com/cms-sw/cmssw/blob/9519f0955cafa3e479b4d085587782d316c5a8cb/RecoVertex/BeamSpotProducer/plugins/BeamSpotOnlineProducer.cc#L91-L93

Reply:
I was unsure about this, if you think it is the proper thing to do I can do that. I did not know what this changeFrame was used for

Reply:
well, if you don't do it, it won't reproduce the previous behaviour, so I think it should be added.

Reply:
done